src/sonic-py-common/sonic_py_common/general.py

```diff
@@ -23,3 +24,60 @@ def load_module_from_source(module_name, file_path):
     sys.modules[module_name] = module
 
     return module
+
+
+def getstatusoutput_noshell(cmd):
+    """
+    This function implements getstatusoutput API from subprocess module
+    but using shell=False to prevent shell injection.
+    """
+    try:
+        output = check_output(cmd, text=True, stderr=STDOUT)
+        exitcode = 0
+    except CalledProcessError as ex:
+        output = ex.output
+        exitcode = ex.returncode
+    if output[-1:] == '\n':
+        output = output[:-1]
+    return exitcode, output
+
+
+def getstatusoutput_noshell_pipe(cmd1, cmd2):
+    """
+    This function implements getstatusoutput API from subprocess module
+    but using shell=False to prevent shell injection. Input command includes
+    pipe command.
+    """
+    cmd = ' '.join(cmd1) + ' | ' + ' '.join(cmd2)
+    try:
+        with Popen(cmd1, universal_newlines=True, stdout=PIPE) as p1:
+            with Popen(cmd2, universal_newlines=True, stdin=p1.stdout, stdout=PIPE) as p2:
+                output = p2.communicate()[0]
+                p1.wait()
+                if p1.returncode != 0 and p2.returncode != 0:
+                    returncode = p1.returncode if p2.returncode == 0 else p2.returncode
+                    raise CalledProcessError(returncode=returncode, cmd=cmd, output=output)
+        exitcode = 0
+    except CalledProcessError as e:
+        output = e.output
+        exitcode = e.returncode
+    if output[-1:] == '\n':
+        output = output[:-1]
+    return exitcode, output
+
+
+def check_output_pipe(cmd1, cmd2):
+    """
+    This function implements check_output API from subprocess module.
+    Input command includes pipe command.
+    """
+    cmd = ' '.join(cmd1) + ' | ' + ' '.join(cmd2)
+    with Popen(cmd1, universal_newlines=True, stdout=PIPE) as p1:
+        with Popen(cmd2, universal_newlines=True, stdin=p1.stdout, stdout=PIPE) as p2:
+            output = p2.communicate()[0]
+            p1.wait()
+            if p1.returncode != 0 and p2.returncode != 0:
+                returncode = p1.returncode if p2.returncode == 0 else p2.returncode
```

Comment:
>p1 [](http://example.com/codeflow?start=29&length=2)

check p1 return code first, if not zero, throw based on p1 info. otherwise, check p2 return code, and throw based on p2 info.